### PR TITLE
coll/spacc: Add Alltoallv implementation

### PR DESCRIPTION
**Pull request by mkurnosov**
1. Add pairwise exchange algorithm for Alltoallv. This algorithm takes into account
   zero values in `scounts[]` and `rcouns[]`.

2. Fix: prepare the placeholder for the array of `request*` (for using linear allreduce).

Signed-off-by: Mikhail Kurnosov <mkurnosov@gmail.com>

### CODE REVIEW

**Comment by bwbarrett** [comment]
ok to test

**Comment by bosilca** [comment]
What is the difference between this and ompi_coll_base_alltoallv_intra_pairwise ?

File: ompi/mca/coll/spacc/coll_spacc_alltoallv.c [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    for (i = 0; i < comm_size; i += block_size) {
+        nreqs = 0;
+        nops = comm_size - i < block_size ? comm_size - i : block_size;
+        /* Post nops irecvs and isends */
+        for (j = 0; j < nops; j++) {
+            remote = (rank + i + j) % comm_size;
+            if (rcounts[remote] <= 0 || rank == remote)
+                continue;
+            precv = ((char *)rbuf) + (ptrdiff_t)rdisps[remote] * rdtype_ext;
+            err = MCA_PML_CALL(irecv_init(precv, rcounts[remote], rdtype,
```
Comment (ggouaillardet) [inline ompi/mca/coll/spacc/coll_spacc_alltoallv.c]: Is there any reason why you prefer `i{send,recv}_init` over `i{send,recv}` ?

If you really want to use `i{send,recv}_init`, then these requests must be properly free'd right after `ompi_request_wait_all()`

**Comment by ggouaillardet** [comment]
@bosilca @mkurnosov 

I think the commit message is misleading since the implemented algorithm is not really a pairwise exchange. (it is more of a multi-pairs wise exchange so to speak).

Also, no message is exchanged when the count is zero.
Note we used to have this implementation (we also handled zero size datatypes) and then we removed it.
Though it might look like a good idea at first glance, it leads to MPI tasks being unsynchronized, and then a lot of unexpected messages that ends up degrading performances. That is why we chose to revert the zero size message optimization.

**Comment by mkurnosov** [comment]
@ggouaillardet @bosilca thank you for your review and comments.

> Note we used to have this implementation (we also handled zero size datatypes) and then we removed it. Though it might look like a good idea at first glance, it leads to MPI tasks being unsynchronized, and then a lot of unexpected messages that ends up degrading performances. That is why we chose to revert the zero size message optimization.

I am agree with your arguments. I think the right solution is to close this PR. Is there list of open tasks (RoadMap) for the ompi/coll/{base,basic,tuned} subsystem?

**Comment by bosilca** [comment]
No at this point there is no common roadmap for any of the collective frameworks.

**Pull request closed without merging**